design/adding_anova_to_mmrm_design_document.qmd

```diff
@@ -0,0 +1,579 @@
+    
+    # Calculate the standard diagnostics.
+    out <-
+      data.frame(
+        call = vapply(fit_calls, deparse1, character(1L)),
+        Model = seq_along(fits),
+        df = vapply(LL_vec, attr, FUN.VALUE = numeric(1L), "df"), 
+        AIC = vapply(fits, AIC, numeric(1)),
+        BIC = vapply(fits, BIC, numeric(1)),
+        logLik = as.numeric(LL_vec)
+      )
+    
+    # If the user has requested the likelihood ratio test...
+    if (test) {
+      # First iterate through and check the models to make sure their covariates
+      # and covariance structures are nested
+      for (i in seq_along(fits)[-1L]) {
+        h_assert_nested_models(fits[[i - 1L]], fits[[i]])
+      }
+      
+      if (!refit && !h_fits_data_identical(fits, refit = FALSE)) {
+        stop("Not all data sets underlying the model fits are identical and ",
+             "refit = FALSE.", call. = FALSE)
+      }
+      
+      # Label for the pair being compared (e.g., "3 vs 4")
+      out$test <- c(NA, paste(head(out$Model, -1), "vs", tail(out$Model, -1)))
+      
+      # Calculate the likelihood ratio between the pairs using base::diff()
+      out$likelihood_ratio <- c(NA, diff(out$logLik))
+      
+      # Calculate the p-value using the ratios and differences in df
+      out$p_value <-
+        pchisq(
+          2 * abs(out$likelihood_ratio),
+          df = abs(diff(out$df)), 
+          lower.tail = FALSE
+        )
+    }
+    
+  } else { 
+    
+    # attributes of terms.object for the model formula without the cov struct
+    terms_attr <- attributes(terms(object$formula_parts$model_formula))
+    
+    # Get a character vector of the terms.
+    # Prepend with "(Intercept)" if an intercept is present
+    terms <- c(if (terms_attr$intercept) "(Intercept)", terms_attr$term.labels)
+    
+    # Vector identifying each coefficient's corresponding term
+    assign_vec <- attr(component(object, "x_matrix"), "assign")
+    
+    # Create an identity matrix with a row/col for each coefficient
+    identity_mtx <- diag(length(assign_vec))
+    
+    # Split the rows of identity_mtx according to assign_vec, creating a list of
+    # contrast matrices: one for each term in the model.
+    contrasts <-
+      split.data.frame(identity_mtx, factor(assign_vec, labels = terms))
+    
+    # Get df_md() results for each contrast matrix. Result is a list of lists
+    df_md_results <- lapply(contrasts, df_md, object = object)
+    
+    # Turn results into a data frame. The row.names identify the terms.
+    out <- do.call(rbind, lapply(df_md_results, as.data.frame))
+  }
+  
+  out
+}
+```
+
+#### Supporting functions:
+
+```{r}
+# Compares the data sets underlying a list of fits. 
+
+# If refit = FALSE, this is a predicate function indicating whether all the data
+# sets are the same.
+
+# If refit = TRUE, all data sets are harmonized, the models are refit using the
+# harmonized data set, and the refits are returned.
```

Comment:
It adds complexity that this function returns different things depending on one argument. Can you refactor such that we have the overall flow:

1) `h_check_same_data()` - whether the data sets are the same or not
2) if not the same and not refit -> error
3) if not the same and refit ->
 a) `h_smallest_common_data()` returns smallest common data set included in all fits (and not the largest data set e.g.)
b) `h_refit_on_data()` refits all models on this smallest common data set
4) if the same -> all good